versioned_docs/version-2021fa/lecture9.md

```diff
@@ -0,0 +1,546 @@
+  - Note: you should always commit and push `yarn.lock` and `package-lock.json`
+- Reduces redundancy in packages
+- Including workspaces as dependencies makes it easier to test your utility functions locally before you publish them online
+
+### Yarn workspaces setup
+
+We will be showing Yarn workspaces setup on our example Filterable Product Table App we've been working on throughout the semester. After lecture 8, we bridged the frontend and the backend together to fetch data from our API endpoints defined in `backend/index.ts` to show in the frontend UI.
+
+We kept these two folders separate and we showed the old way of `cd` into each folder and running Yarn install. Now with workspaces we can manage the package from the root and directly run `yarn install` once to install all frontend and backend dependencies!
+
+### Yarn workspaces setup
+
+We will be showing Yarn workspaces setup on our example Filterable Product Table App we've been working on throughout the semester. After lecture 8, we bridged the frontend and the backend together to fetch data from our API endpoints defined in `backend/index.ts` to show in the frontend UI.
+
+We kept these two folders separate and we showed the old way of `cd` into each folder and running Yarn install. Now with workspaces we can manage the package from the root and directly run `yarn install` once to install all frontend and backend dependencies!
+
+#### Root
+
+To do this, our root `package.json` will define the workspaces. (Note: root means outside of both the frontend and backend folders, on that same level of the directory hierarchy.)
+
+Most important lines here are 3-7 which define the workspaces and set `private: true` since [workspaces can't be published](https://classic.yarnpkg.com/en/docs/workspaces/#toc-how-to-use-it).
+
+```json title="/package.json"
+{
+  "name": "products-app",
+  "private": true,
+  "workspaces": ["backend", "frontend"],
+  "version": "1.0.0",
+  "main": "index.js",
+  "license": "MIT"
+}
+```
+
+#### Backend
+
+Now our backend workspace will look like the following. I've omitted some of the dependencies part for brevity you can find the full file [below](./lecture9#backendpackagejson)
+
+```json title="backend/package.json"
+{
+  "name": "backend",
+  "version": "1.0.0",
+  "main": "index.js",
+  "scripts": {
+    "build": "tsc -p tsconfig.json",
+    "start": "node index.js"
+  },
+  "dependencies": {
+    "cors": "^2.8.5",
+    "express": "^4.17.1",
+    "firebase-admin": "^9.4.1"
+  }
+  // dev dependencies...
+}
+```
+
+#### Frontend
+
+Our frontend workspace will also look like the following. The full file can also be found is very similar to the default `package.json` from create react app we have been using all along.
+
+```json title="frontend/package.json"
+{
+  "name": "frontend",
+  "version": "0.1.0",
+  "private": true,
+  "proxy": "http://localhost:8080",
+  "dependencies": {
```

Comment:
same with here, especially b/c Firebase v8 is different from Firebase v9 (the one used in demo)